### PR TITLE
Trying to fix Two Definitions error

### PR DESCRIPTION
**Pull request by EIIKaO**
#### Summary
Bugfixes "fix:Unexpected two definitions from the same source (dda) error"

#### Purpose of change
Trying to fix [#59508](https://github.com/CleverRaven/Cataclysm-DDA/issues/59508#issuecomment-1193070350)

bug reproduction method
1. Copy attached modinfo.json([modinfo.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/9177432/modinfo.txt) **rename ext from txt to json**) into your "data/mods/anyname_or_two_difinitions" folder.
2. Run C:DDA, and Json error raised. But it is harmless at this stage. Ignore it.
3. Create a new world and include the "RSM_X04" mod(atached file) in its mods list.
4. Start a new game in the world you created above.
5. Json error raised again. New game is abort.
6. Ok, create another new world and **exclude** the "RSM_X04" mod(attached file) in its mods list.
7. Start a new game in the world you created the last.
8. Raise "Error: ***** (*****) has two definitions from the same source (dda)!" and new game is aborting again(and you can't play anymore).

Reboot C:DDA and start a new game in the world you created the last(or other world that bad mods exclued ), then no error and you can play game. Now trash the attached modinfo.json.

#### Describe the solution
It seems some of DynamicDataLoader::deferred_json objects were not cleared(or not working as expected.).
I just inserted the DynamicDataLoader::deferred_json.clear() method.
Please point out if my perception is incorrect.

By code changing, at least the unexpected "two definitions error" is no longer raised in my situation.
And legitimate "two definition errors" are still arise without problems(as designed).

#### Describe alternatives you've considered

I am not a C++ expert (especially about modern programming).
So I did not traced why the list was not cleared in current code.

Also I think my approach could cause the following problems.
1. Reduce performance(deferred_json is may related to cache?)
2. Memory leak(I am clearing the list without a good ponder over)

Code changes are small, but if there are problems with the code, the impact could be large.
I hope for advice and rewrite from C++ experts and veteran C:DDA developer.

#### Testing
I have tried new and load game several times.(I believe the impact of the code change would be limited to the start of the new game and load game).

#### Additional context
None

### CODE REVIEW

**Comment by anothersimulacrum** [comment]
Thank you so much for finding this, I'll give it a look.

**Review by anothersimulacrum — APPROVED** [review]
tl:dr; This fixes the issue, and is the correct solution.

Okay, so `DynamicDataLoader::load_deferred()` is supposed to clear the deferred JSON it loads, and is called in the finalization stage of loading JSON. Except, when we throw an error during loading, we never reach that stage, and so it is never cleared.

Okay, so this seems correct - let's look at the potential problems:

> Reduce performance(deferred_json is may related to cache?)

Nope, it's JSON data we have yet to load due to inheritance, there's no caching, and it definitely should be cleared before we start trying to load again.

> Memory leak(I am clearing the list without a good ponder over)

`deferred` is a `DynamicDataLoader:deferred_json` which is an `std::list<std::pair<json_source_location, std::string>>`. A `json_source_location` consists of a `shared_pointer` to an `std::string` filename, and an `int` offset into the file. All of this is safe to clear, no manual cleanup needed.
